### PR TITLE
Improving PHPStan configuration

### PR DESCRIPTION
**Pull request by ondrejmirtes**

### CODE REVIEW

File: src/Shell/I18nShell.php
Comment (dereuromark) [inline src/Shell/I18nShell.php]: protected

File: src/TestSuite/LegacyRequestDispatcher.php
Comment (dereuromark) [inline src/TestSuite/LegacyRequestDispatcher.php]: also protected

File: src/Shell/PluginShell.php
Comment (ADmad) [inline src/Shell/PluginShell.php]: `@property \Cake\Shell\Task\AssetsTask $Assets`

Similarly the other 2 properties belows are also tasks from `\Cake\Shell\Task\` namespace.

**Review by ADmad — CHANGES_REQUESTED** [review]

**Comment by ondrejmirtes** [comment]
Please don't merge it right away :) I'd like to bring the number of errors to as low as possible.

**Comment by ADmad** [comment]
@ondrejmirtes I have a couple handy for you :) https://gist.github.com/ADmad/ca6627799292d155d36e6555e7b4a8f4

**Comment by ondrejmirtes** [comment]
I think you can  commit and push to my branch:

```
git push git@github.com:ondrejmirtes/cakephp.git HEAD:phpstan
```

**Comment by ADmad** [comment]
I believe that would require you to give me commit rights to your fork. It's no biggie, I cam make those changes later on our repo.

**Comment by ondrejmirtes** [comment]
I pushed a custom class reflection extension to cover the Association/Table mixin behaviour, to prevent these errors:

```
------ ------------------------------------------------------------------- 
  Line   src/ORM/Association/HasOne.php                                     
 ------ ------------------------------------------------------------------- 
  153    Call to an undefined method Cake\ORM\Association\HasOne::alias().  
 ------ ------------------------------------------------------------------- 
 ------ ---------------------------------------------------------------------- 
  Line   src/ORM/Association/BelongsTo.php                                     
 ------ ---------------------------------------------------------------------- 
  208    Call to an undefined method Cake\ORM\Association\BelongsTo::alias().  
 ------ ---------------------------------------------------------------------- 
 ------ ------------------------------------------------------------------------ 
  Line   src/ORM/Association/Loader/SelectWithPivotLoader.php                    
 ------ ------------------------------------------------------------------------ 
  174    Access to an undefined property                                         
         Cake\ORM\Association\Loader\SelectWithPivotLoader::$_junctionProperty.  
 ------ ------------------------------------------------------------------------ 
 ------ ---------------------------------------------- 
  Line   src/ORM/Association/BelongsToMany.php         
 ------ ---------------------------------------------- 
  550    Call to an undefined method                   
         Cake\ORM\Association\BelongsToMany::alias().  
  1027   Call to an undefined method                   
         Cake\ORM\Association\BelongsToMany::alias().  
 ------ ---------------------------------------------- 
 ------ -------------------------------------------------------------------- 
  Line   src/ORM/Association/HasMany.php                                     
 ------ -------------------------------------------------------------------- 
  658    Call to an undefined method Cake\ORM\Association\HasMany::alias().  
 ------ -------------------------------------------------------------------- 
```

PHPStan now knows that methods in Association children can come from the Table class and checks correctly their usage. Do you like it?

File: config/PHPStan/AssociationTableMixinClassReflectionExtension.php
Comment (dereuromark) [inline config/PHPStan/AssociationTableMixinClassReflectionExtension.php]: We dont need the underscores here as prefix.

File: src/TestSuite/LegacyRequestDispatcher.php
Comment (dereuromark) [inline src/TestSuite/LegacyRequestDispatcher.php]: unfolded maybe?

File: config/PHPStan/AssociationTableMixinClassReflectionExtension.php
Comment (ondrejmirtes) [inline config/PHPStan/AssociationTableMixinClassReflectionExtension.php]: stickler-ci complained about it

File: src/TestSuite/LegacyRequestDispatcher.php
Comment (ondrejmirtes) [inline src/TestSuite/LegacyRequestDispatcher.php]: I don't understand?

Comment (dereuromark) [inline src/TestSuite/LegacyRequestDispatcher.php]: Usually we use multi line for those. This is a single line annotation,

File: config/PHPStan/AssociationTableMixinClassReflectionExtension.php
Comment (dereuromark) [inline config/PHPStan/AssociationTableMixinClassReflectionExtension.php]: Ah ok, but _ should be protected then.

**Comment by ADmad** [comment]
You'll need to update this test case https://travis-ci.org/cakephp/cakephp/jobs/187818797#L698

**Comment by ondrejmirtes** [comment]
@ADmad I meant to ask whether you think the `config/PHPStan` is the right directory for this kind of stuff? I'd much rather add a new directory `build` that sets a more clear purpose.

**Comment by dereuromark** [comment]
I think it is fine as it is.

**Comment by ADmad** [comment]
I think having the phpstan files inside `tests/` would be better.

**Comment by ADmad** [comment]
> Do you consider removing the build from allow_failures when we get the number of errors to zero?

Yes I would.

File: tests/TestCase/Filesystem/FolderTest.php
Comment (ADmad) [inline tests/TestCase/Filesystem/FolderTest.php]: Please use `DS` instead of `/` else the test will fail on windows.

File: tests/TestCase/Filesystem/FolderTest.php
Comment (ADmad) [inline tests/TestCase/Filesystem/FolderTest.php]: Same as above, use `DS` instead of `/` else the test will fail on windows.

**Review by ADmad — CHANGES_REQUESTED** [review]

**Comment by ADmad** [comment]
@ondrejmirtes This seems good to merge once the phpstan files are moved from `/config` to `/tests`.

You can open new PR with further fixes. Incremental changes will be easier to review.

**Comment by ondrejmirtes** [comment]
I pushed a new version, let's see if it builds :)

**Comment by ADmad** [comment]
Getting there: "Found 27 errors".

I am merging this to avoid more merge conflicts.

**Pull request merged**
